sync/src/relayer/compact_block_process.rs

```diff
@@ -273,3 +160,175 @@ impl<'a> HeaderProvider for CompactBlockMedianTimeView<'a> {
+        return StatusCode::CompactBlockRequiresParent.with_context(format!(
+            "{} parent: {}",
+            block_hash,
+            compact_block_header.data().raw().parent_hash(),
+        ));
+    }
+
+    // compact block is in pending
+    let pending_compact_blocks = shared.state().pending_compact_blocks();
+    if pending_compact_blocks
+        .get(&block_hash)
+        .map(|(_, peers_map, _)| peers_map.contains_key(&peer))
+        .unwrap_or(false)
+    {
+        return StatusCode::CompactBlockIsAlreadyPending.with_context(block_hash);
+    }
+
+    // compact header verification
+    let fn_get_pending_header = {
+        |block_hash| {
+            pending_compact_blocks
+                .get(&block_hash)
+                .map(|(compact_block, _, _)| compact_block.header().into_view())
+                .or_else(|| {
+                    shared
+                        .get_header_view(&block_hash, None)
+                        .map(|header_view| header_view.into_inner())
+                })
+        }
+    };
+    let median_time_context = CompactBlockMedianTimeView {
+        fn_get_pending_header: Box::new(fn_get_pending_header),
+    };
+    let header_verifier = HeaderVerifier::new(&median_time_context, shared.consensus());
+    if let Err(err) = header_verifier.verify(compact_block_header) {
+        if err
+            .downcast_ref::<HeaderError>()
+            .map(|e| e.is_too_new())
+            .unwrap_or(false)
+        {
+            return Status::ignored();
+        } else {
+            shared
+                .state()
+                .insert_block_status(block_hash.clone(), BlockStatus::BLOCK_INVALID);
+            return StatusCode::CompactBlockHasInvalidHeader
+                .with_context(format!("{} {}", block_hash, err));
+        }
+    }
+
+    Status::ok()
+}
+
+/// request missing txs and uncles from peer
+fn missing_or_collided_post_process(
+    compact_block: CompactBlock,
+    block_hash: Byte32,
+    mut pending_compact_blocks: MutexGuard<PendingCompactBlockMap>,
+    nc: Arc<dyn CKBProtocolContext>,
+    missing_transactions: Vec<u32>,
+    missing_uncles: Vec<u32>,
+    peer: PeerIndex,
+) {
+    pending_compact_blocks
+        .entry(block_hash.clone())
+        .or_insert_with(|| (compact_block, HashMap::default(), unix_time_as_millis()))
+        .1
+        .insert(peer, (missing_transactions.clone(), missing_uncles.clone()));
+
+    let content = packed::GetBlockTransactions::new_builder()
+        .block_hash(block_hash)
+        .indexes(missing_transactions.pack())
+        .uncle_indexes(missing_uncles.pack())
+        .build();
+    let message = packed::RelayMessage::new_builder().set(content).build();
+    let sending = send_message_to(nc.as_ref(), peer, &message);
+    if !sending.is_ok() {
+        ckb_logger::warn_target!(
+            crate::LOG_TARGET_RELAY,
+            "ignore the sending message error, error: {}",
+            sending
+        );
+    }
```

Comment:
let content = packed::GetBlockTransactions::new_builder()
        .block_hash(block_hash)
        .indexes(missing_transactions.pack())
        .uncle_indexes(missing_uncles.pack())
        .build();
    let message = packed::RelayMessage::new_builder().set(content).build();
    match send_message_to(nc.as_ref(), peer, &message) {
        Ok(_) => {
            pending_compact_blocks
                .entry(block_hash.clone())
                .or_insert_with(|| (compact_block, HashMap::default(), unix_time_as_millis()))
                .1
                .insert(peer, (missing_transactions, missing_uncles));
        }
        Err(err) => {
            ckb_logger::warn_target!(
                crate::LOG_TARGET_RELAY,
                "ignore the sending message error, error: {:?}",
                err
            );
        }
    }